File: project_euler/problem_479/solution.py
"""
Problem 479: Roots on the Rise

S(n) = sum_{k=1}^{n} sum_{p=1}^{n} (1-k^2)^p  mod 10^9+7

Key insight: (a_k+b_k)(b_k+c_k)(c_k+a_k) = 1 - k^2 from Vieta's formulas.

For each k >= 2, the inner sum is a geometric series.
For k = 1, 1-k^2 = 0 so contribution is 0.
"""

MOD = 1000000007

def solve(n):
    ans = 0

    for k in range(2, n + 1):
        k2 = k * k % MOD
        q = (1 - k2) % MOD  # q = 1 - k^2 mod MOD

        # sum_{p=1}^{n} q^p = (q^{n+1} - q) / (q - 1)
        # q - 1 = -k^2
        # So sum = (q^{n+1} - q) / (-k^2) = -(q^{n+1} - q) * inv(k^2)

        qn1 = pow(q, n + 1, MOD)
        num = (qn1 - q) % MOD

        inv_k2 = pow(k2, MOD - 2, MOD)
        term = (-num * inv_k2) % MOD

        ans = (ans + term) % MOD

    return ans % MOD

if __name__ == "__main__":
    # Verify S(4) = 51160
    print(f"S(4) = {solve(4)}")

    # Compute S(10^6)
    result = solve(1000000)
    print(f"S(10^6) mod 10^9+7 = {result}")

File: project_euler/problem_479/solution.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
const ll MOD = 1000000007LL;

ll power(ll base, ll exp, ll mod) {
    ll result = 1;
    base %= mod;
    if (base < 0) base += mod;
    while (exp > 0) {
        if (exp & 1) result = result * base % mod;
        base = base * base % mod;
        exp >>= 1;
    }
    return result;
}

ll modinv(ll a, ll mod) {
    return power(a, mod - 2, mod);
}

int main() {
    ll n = 1000000; // 10^6

    ll ans = 0;

    // For k=1: 1 - k^2 = 0, so contribution is 0
    // For k >= 2:
    // sum_{p=1}^{n} (1-k^2)^p = ((1-k^2)^{n+1} - (1-k^2)) / ((1-k^2) - 1)
    //                          = ((1-k^2)^{n+1} - (1-k^2)) / (-k^2)

    for (ll k = 2; k <= n; k++) {
        ll k2 = k % MOD * (k % MOD) % MOD;
        ll q = (1 - k2 % MOD + MOD) % MOD; // q = 1 - k^2 mod MOD

        // Numerator: q^{n+1} - q
        ll qn1 = power(q, n + 1, MOD);
        ll num = (qn1 - q + MOD) % MOD;

        // Denominator: q - 1 = -k^2
        // So we divide by (q - 1) = -k^2
        // num / (q - 1) = num / (-k^2) = -num * inv(k^2)
        ll inv_k2 = modinv(k2, MOD);
        ll term = (MOD - num) % MOD * inv_k2 % MOD;

        ans = (ans + term) % MOD;
    }

    printf("%lld\n", ans);

    return 0;
}
